### PR TITLE
Add policies for handling particles in guard cells

### PR DESCRIPTION
**Pull request by Flamefire**
This allows to change what happens with particles in guard cells.

Could help #123 

Idea was discussed with @ax3l and this is preferred over e.g. virtual function calls

### CODE REVIEW

File: src/libPMacc/include/particles/ParticleDescription.hpp [hunk 1]
```diff
@@ -55,7 +56,8 @@ typename T_SuperCellSize,
 typename T_ValueTypeSeq,
 typename T_Flags = bmpl::vector0<>,
 typename T_MethodsList = bmpl::vector0<>,
-typename T_FrameExtensionList = bmpl::vector0<>
+typename T_FrameExtensionList = bmpl::vector0<>,
+typename T_HandleGuardParticles = particles::HandleGuardParticles<>
```
Comment (Flamefire) [inline src/libPMacc/include/particles/ParticleDescription.hpp]: Move 2 up (as discussed with @psychocoderHPC )

**Comment by Flamefire** [comment]
d8f95a3 makes the handling more generic and less error-prone

File: src/libPMacc/include/particles/DeleteParticles.hpp [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+#pragma once
+
+namespace PMacc {
+namespace particles {
+
+    /**
+     * Policy for HandleGuardParticles that removes all particles from guard cells
+     */
+    struct DeleteParticles
```
Comment (psychocoderHPC) [inline src/libPMacc/include/particles/DeleteParticles.hpp]: Can you please add a new folder in `/include/particles/` named `policies` or `policy`. Please don't forget to add the namespace too.

File: src/libPMacc/include/particles/tasks/TaskParticlesReceive.hpp [hunk 3]
```diff
@@ -49,15 +57,25 @@ namespace PMacc
...
             EventTask serialEvent = __getTransactionEvent();
+            HandleExchanged handleExchanged;
+            HandleNotExchanged handleNotExchanged;
 
             for (int i = 1; i < Exchanges; ++i)
             {
-                if (parBase.getParticlesBuffer().hasReceiveExchange(i))
-                {
+                /* Start new transaction */
+                if(canRunInParallel)
```
Comment (psychocoderHPC) [inline src/libPMacc/include/particles/tasks/TaskParticlesReceive.hpp]: Is this really needed or is this a micro-optimization?

Comment (psychocoderHPC) [inline src/libPMacc/include/particles/tasks/TaskParticlesReceive.hpp]: Please remove `if(canRunInParallel)`, to accept this we need some race condition checks with PIConGPU. Thus this is time intensive we should use the known version here.

File: src/libPMacc/include/particles/tasks/TaskParticlesReceive.hpp [hunk 4]
```diff
@@ -49,15 +57,25 @@ namespace PMacc
...
                     __startAtomicTransaction(serialEvent);
-                    Environment<>::get().ParticleFactory().createTaskReceiveParticlesExchange(parBase, i);
-                    tmpEvent += __endTransaction();
-                }
+
+                /* Handle particles */
+                if (parBase.getParticlesBuffer().hasReceiveExchange(i))
+                    handleExchanged.handleOutgoing(parBase, i);
+                else
+                    handleNotExchanged.handleOutgoing(parBase, i);
```
Comment (psychocoderHPC) [inline src/libPMacc/include/particles/tasks/TaskParticlesReceive.hpp]: this should be `handleIncoming`

File: src/libPMacc/include/particles/tasks/TaskParticlesReceive.hpp [hunk 5]
```diff
@@ -49,15 +57,25 @@ namespace PMacc
...
+                    __startTransaction(serialEvent);
+                else
                     __startAtomicTransaction(serialEvent);
-                    Environment<>::get().ParticleFactory().createTaskReceiveParticlesExchange(parBase, i);
-                    tmpEvent += __endTransaction();
-                }
+
+                /* Handle particles */
+                if (parBase.getParticlesBuffer().hasReceiveExchange(i))
+                    handleExchanged.handleOutgoing(parBase, i);
```
Comment (psychocoderHPC) [inline src/libPMacc/include/particles/tasks/TaskParticlesReceive.hpp]: this should be `handleIncoming`

**Comment by psychocoderHPC** [comment]
For this pull request we need a KHI heating test!

- [ ] KHI heating test

File: src/libPMacc/include/particles/DeleteParticles.hpp (hunk 2 above)
Comment (Flamefire) [inline src/libPMacc/include/particles/DeleteParticles.hpp]: Done

File: src/libPMacc/include/particles/tasks/TaskParticlesReceive.hpp (hunk 3 above)
Comment (Flamefire) [inline src/libPMacc/include/particles/tasks/TaskParticlesReceive.hpp]: The currently used policies set this to false. So the "known version" is used here. If the policies only read the frames, __startTransaction would be possible here. But currenlty it is not used, but only a preparation for future implementations.

File: src/libPMacc/include/particles/tasks/TaskParticlesReceive.hpp (hunk 5 above)
Comment (Flamefire) [inline src/libPMacc/include/particles/tasks/TaskParticlesReceive.hpp]: Done

File: src/libPMacc/include/particles/tasks/TaskParticlesReceive.hpp (hunk 4 above)
Comment (Flamefire) [inline src/libPMacc/include/particles/tasks/TaskParticlesReceive.hpp]: Done

**Comment by Flamefire** [comment]
KHI heating is same as in dev

**Comment by Flamefire** [comment]
Checks succeeded also with new commit

**Pull request merged**